### PR TITLE
enhancement(nats sink): Confrom error events to spec, add dropped events

### LINKED ISSUE #14019: Verify `nats` sink
- [x] Verify internal event processing metrics are tested via `assert_X_compliance` test helpers
- [x] Update `component_discarded_events_total` to match the new [instrumentation spec](https://github.com/vectordotdev/vector/blob/master/docs/specs/instrumentation.md#eventsdropped) by adding `reason` to the emitted logs, and `intentional` to emitted logs and metrics
- [x] Verify that the component correctly emits the correct error events per the [component spec](https://github.com/vectordotdev/vector/blob/master/docs/specs/instrumentation.md#events)

### PR DESCRIPTION
**Pull request by spencergilbert**
Signed-off-by: Spencer Gilbert <spencer.gilbert@datadoghq.com>

- Adds rate limit to SendError event.
- Adds EventsDropped to SendError as we don't know if source will retry.

Closes #14019

### CODE REVIEW

File: src/internal_events/nats.rs [hunk 1]
```diff
@@ -27,5 +30,10 @@ impl InternalEvent for NatsEventSendError {
         );
         // deprecated
         counter!("send_errors_total", 1);
+        emit!(ComponentEventsDropped {
+            count: 1,
+            intentional: false,
+            reason: reason,
```
Comment (neuronull) [inline src/internal_events/nats.rs:36]: I think this will work:
```suggestion
            reason,
```

**Review by neuronull — APPROVED** [review]

File: src/internal_events/nats.rs (hunk 1 above)
Comment (spencergilbert) [reply src/internal_events/nats.rs:36]: Huh, yeah I think so too - I expected rust-analyzer to ding me there though. Double checking before committing the suggestion.

File: src/internal_events/nats.rs [hunk 2]
```diff
@@ -27,5 +30,10 @@ impl InternalEvent for NatsEventSendError {
         );
         // deprecated
```
Comment (jszwedko) [inline src/internal_events/nats.rs:31]: Small thing, but this `//deprecated` is a bit ambiguous as to whether it applies to the next line or next block. I think some additional whitespace (or moving this down to the bottom) would help.

**Review by jszwedko — APPROVED** [review]

File: src/internal_events/nats.rs (hunk 2 above)
Comment (spencergilbert) [reply src/internal_events/nats.rs:31]: True, I had thought that briefly too but ignored the feeling.

Comment (spencergilbert) [reply src/internal_events/nats.rs]: I dropped the deprecated one to the bottom separated by a newline, [38aa514](https://github.com/vectordotdev/vector/pull/14217/commits/38aa5148653430f9910905f95fd2bf821178e1a1)

**Pull request merged**